scripts/ingest-tarball.sh

```diff
@@ -24,11 +31,151 @@ function echo_red() {
+    fi
+
+    # Check if the second-level dir in the tarball is compat, software, or init
+    if [ ! -v content_types[${tar_contents_type_dir}] ]
+    then
+        error "the second directory level of the tarball contents should be either compat, software, or init."
+    fi
+}
+
+function cvmfs_regenerate_nested_catalogs() {
+    # Use the .cvmfsdirtab to generate nested catalogs for the ingested tarball
+    # ("cvmfs_server ingest" doesn't do this automatically)
+    echo "Generating the nested catalogs..."
+    cvmfs_server transaction "${repo}"
+    cvmfs_server publish -m "Generate catalogs after ingesting ${tar_file_basename}" "${repo}"
+    ec=$?
+    if [ $ec -eq 0 ]
+    then
+        echo_green "Nested catalogs for ${repo} have been created!"
+    else
+        echo_red "failure when creating nested catalogs for ${repo}."
+    fi
+}
+
+function cvmfs_ingest_tarball() {
+    # Do a regular "cvmfs_server ingest" for a given tarball,
+    # followed by regenerating the nested catalog
+    echo "Ingesting tarball ${tar_file} to ${repo}..."
+    ${decompress} "${tar_file}" | cvmfs_server ingest -t - -b "${basedir}" "${repo}"
+    ec=$?
+    if [ $ec -eq 0 ]
+    then
+        echo_green "${tar_file} has been ingested to ${repo}."
+    else
+        error "${tar_file} could not be ingested to ${repo}."
+    fi
+
+    cvmfs_regenerate_nested_catalogs
+}
+
+function check_os() {
+    # Check if the operating system directory is correctly set for the contents of the tarball
+    os=$(echo "${tar_first_file}" | cut -d / -f 3)
+    if [ -z "${os}" ]
+    then
+        error "no operating system directory found in the tarball!"
+    fi
+    if [ ! -v oss[${os}] ]
+    then
+        error "the operating system directory in the tarball is ${os}, which is not a valid operating system!"
+    fi
+}
+
+function check_arch() {
+    # Check if the architecture directory is correctly set for the contents of the tarball
+    arch=$(echo "${tar_first_file}" | cut -d / -f 4)
+    if [ -z "${arch}" ]
+    then
+        error "no architecture directory found in the tarball!"
+    fi
+    if [ ! -v archs[${arch}] ]
+    then
+        error "the architecture directory in the tarball is ${arch}, which is not a valid architecture!"
+    fi
+}
+
+function ingest_init_tarball() {
+    # Handle the ingestion of tarballs containing init scripts
+    cvmfs_ingest_tarball
+}
+
+function ingest_software_tarball() {
+    # Handle the ingestion of tarballs containing software installations
+    check_arch
+    check_os
+    cvmfs_ingest_tarball
+}
+
+function ingest_compat_tarball() {
+    # Handle the ingestion of tarballs containing a compatibility layer
+    check_arch
+    check_os
+    # Assume that we already had a compat layer in place if there is a startprefix script in the corresponding CVMFS directory
+    if [ -f "/cvmfs/${repo}/${basedir}/${version}/compat/${os}/${arch}/startprefix" ];
```

Comment:
What if `${repo}` is not defined? Same for `${version}`, `${basedir}` (and actually also for `${os}` and `${arch}`, although with the current implementation of `check_arch` they won't be, but I'm not sure we should rely on that)

Reply:
Because of the added checks for `repo, version, basedir`, these shouldn't be undefined anymore here. But I don't understand why we wouldn't be able to rely on `check_arch` here?

Reply:
It's mostly about not making assumptions about which variables get set by other functions, but perhaps I'm a bit too pedantic ;)